File: api/src/main/java/fr/huiitre/tools/modules/riot/valorant/api/ValorantWatchlistController.java
package fr.huiitre.tools.modules.riot.valorant.api;

import fr.huiitre.tools.modules.core.common.api.RequiredRole;
import fr.huiitre.tools.modules.core.role.domain.RoleCode;
import fr.huiitre.tools.modules.riot.valorant.application.user.command.AddToWatchlistCommand;
import fr.huiitre.tools.modules.riot.valorant.application.user.usecase.AddSkinToWatchlistUseCase;
import fr.huiitre.tools.modules.riot.valorant.application.user.usecase.GetMyValorantWatchlistUseCase;
import fr.huiitre.tools.modules.riot.valorant.application.user.usecase.RemoveSkinFromWatchlistUseCase;
import fr.huiitre.tools.modules.riot.valorant.application.user.usecase.ValorantWatchlistNotifier;
import fr.huiitre.tools.modules.riot.valorant.application.skin.view.ValorantSkinView;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/riot/valorant/watchlist")
public class ValorantWatchlistController {

    private final GetMyValorantWatchlistUseCase getMyWatchlistUseCase;
    private final AddSkinToWatchlistUseCase addSkinToWatchlistUseCase;
    private final RemoveSkinFromWatchlistUseCase removeSkinFromWatchlistUseCase;
    private final ValorantWatchlistNotifier watchlistNotifier;

    public ValorantWatchlistController(
            GetMyValorantWatchlistUseCase getMyWatchlistUseCase,
            AddSkinToWatchlistUseCase addSkinToWatchlistUseCase,
            RemoveSkinFromWatchlistUseCase removeSkinFromWatchlistUseCase,
            ValorantWatchlistNotifier watchlistNotifier) {
        this.getMyWatchlistUseCase = getMyWatchlistUseCase;
        this.addSkinToWatchlistUseCase = addSkinToWatchlistUseCase;
        this.removeSkinFromWatchlistUseCase = removeSkinFromWatchlistUseCase;
        this.watchlistNotifier = watchlistNotifier;
    }

    @RequiredRole(RoleCode.READ_ONLY)
    @GetMapping
    public List<ValorantSkinView> getMyWatchlist() {
        return getMyWatchlistUseCase.execute();
    }

    @RequiredRole(RoleCode.USER)
    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    public ValorantSkinView addToWatchlist(@RequestBody AddToWatchlistCommand command) {
        return addSkinToWatchlistUseCase.execute(command);
    }

    @RequiredRole(RoleCode.USER)
    @DeleteMapping("/{skinId}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void removeFromWatchlist(@PathVariable Long skinId) {
        removeSkinFromWatchlistUseCase.execute(skinId);
    }

    @RequiredRole(RoleCode.ADMIN)
    @PostMapping("/admin/sync")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void triggerSync() {
        watchlistNotifier.processAllUsers();
    }
}

File: api/Modules/Riot/Valorant/Api/ValorantWatchlistController.cs
using Microsoft.AspNetCore.Mvc;
using Tools.Api.Modules.Riot.Valorant.Application.Skin.Views;
using Tools.Api.Modules.Riot.Valorant.Application.User.Commands;
using Tools.Api.Modules.Riot.Valorant.Application.User.Usecases;

namespace Tools.Api.Modules.Riot.Valorant.Api;

// Skins suivis, par compte Valorant lié.
[ApiController]
[Route("riot/valorant/watchlist")]
public class ValorantWatchlistController : ControllerBase
{
    [HttpGet]
    public Task<List<ValorantSkinView>> GetMyWatchlist(
        [FromServices] GetMyValorantWatchlistUseCase getMyValorantWatchlistUseCase,
        [FromQuery] long accountId
    )
    {
        return getMyValorantWatchlistUseCase.Execute(accountId);
    }

    [HttpPost]
    public async Task<IActionResult> AddToWatchlist(
        [FromServices] AddSkinToWatchlistUseCase addSkinToWatchlistUseCase,
        [FromBody] AddToWatchlistCommand command
    )
    {
        var skin = await addSkinToWatchlistUseCase.Execute(command);

        return StatusCode(StatusCodes.Status201Created, skin);
    }

    [HttpDelete("{skinId:long}")]
    public async Task<IActionResult> RemoveFromWatchlist(
        [FromServices] RemoveSkinFromWatchlistUseCase removeSkinFromWatchlistUseCase,
        [FromRoute] long skinId,
        [FromQuery] long accountId
    )
    {
        await removeSkinFromWatchlistUseCase.Execute(skinId, accountId);

        return NoContent();
    }

    // Déclenche à la main la passe que le planificateur fait seul. Le contrôle d'accès est dans
    // le use case, pas dans le notifieur qu'il appelle — celui-ci tourne aussi sans utilisateur.
    [HttpPost("admin/sync")]
    public async Task<IActionResult> TriggerSync(
        [FromServices] TriggerValorantWatchlistSyncUseCase triggerValorantWatchlistSyncUseCase
    )
    {
        await triggerValorantWatchlistSyncUseCase.Execute();

        return NoContent();
    }
}
